tests/http_spec.rs

```diff
@@ -0,0 +1,216 @@
+use std::collections::BTreeMap;
+use std::fs;
+use std::path::Path;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_graphql_value::ConstValue;
+use hyper::{Body, Request};
+use reqwest::header::{HeaderName, HeaderValue};
+use serde::Deserialize;
+use serde_json::Value;
+use tailcall::blueprint::Blueprint;
+use tailcall::config::Config;
+use tailcall::http::{graphql_request, HttpClient, Method, Response, ServerContext};
+use url::Url;
+
+#[derive(Deserialize, Clone, Debug)]
+pub enum Annotation {
+  Skip,
+  Only,
+  Fail,
+}
+#[derive(Deserialize, Clone, Debug, Eq, PartialEq)]
+pub struct APIRequest {
+  #[serde(default)]
+  method: Method,
+  pub url: Url,
+  #[serde(default)]
+  pub headers: BTreeMap<String, String>,
+  #[serde(default)]
+  pub body: serde_json::Value,
+}
+#[derive(Deserialize, Clone, Debug)]
+pub struct APIResponse {
+  pub status: u16,
+  #[serde(default)]
+  pub headers: BTreeMap<String, String>,
+  #[serde(default)]
+  pub body: serde_json::Value,
+}
+#[derive(Deserialize, Clone, Debug, Eq, PartialEq)]
+pub struct UpstreamRequest(pub APIRequest);
+#[derive(Deserialize, Clone, Debug)]
+pub struct UpstreamResponse(APIResponse);
+#[derive(Deserialize, Clone)]
+pub struct DownstreamRequest(pub APIRequest);
+#[derive(Deserialize, Clone)]
+pub struct DownstreamResponse(pub APIResponse);
+#[derive(Deserialize, Clone)]
+pub struct DownstreamAssertion {
+  pub request: DownstreamRequest,
+  pub response: DownstreamResponse,
+  pub annotation: Option<Annotation>,
+}
+
+#[derive(Default, Deserialize, Clone)]
+pub struct HttpSpec {
+  pub config: String,
+  pub name: String,
+  pub description: Option<String>,
+  pub upstream_mocks: Vec<(UpstreamRequest, UpstreamResponse)>,
+  pub expected_upstream_requests: Vec<UpstreamRequest>,
+  pub downstream_assertions: Vec<DownstreamAssertion>,
+}
+
+impl HttpSpec {
+  fn read<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
+    let path = path.as_ref();
+    let contents = fs::read_to_string(path)?;
+    let extension = path
+      .extension()
+      .ok_or(anyhow!("not a valid extension"))?
+      .to_str()
+      .ok_or(anyhow!("not a valid Unicode"))?;
+
+    let spec = match extension.to_lowercase().as_str() {
+      "json" => anyhow::Ok(serde_json::from_str(&contents)?),
+      "yaml" => anyhow::Ok(serde_yaml::from_str(&contents)?),
+      _ => Err(anyhow!("only json and yaml are supported")),
+    };
```

Comment:
Can leverage `source.rs` instead of duplicating code.

Reply:
done